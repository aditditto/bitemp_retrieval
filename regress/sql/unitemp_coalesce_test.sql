DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'employee',
	$$emp_id int, 
	  emp_name text not null,
    dept text not null
	$$,
   'emp_id');

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$1, 'Ron', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-01'::timestamptz, '2010-10-06'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$2, 'George', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-05'::timestamptz, '2010-10-09'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$1, 'Ron', 'Mail'$$,
    temporal_relationships.timeperiod('2010-10-06'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);


SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Mail'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-08'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Ship'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-10'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Ship'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-15'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);


SELECT emp_id, emp_name, dept, effective FROM employee;

SELECT emp_id, emp_name, dept, effective FROM employee WHERE now() <@ asserted ORDER BY emp_name, asserted;

SELECT * FROM unitemp_coalesce_select(
    'SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted'::text, 
    ARRAY['emp_name', 'dept'], 
    'effective'::text
    ) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;

SELECT * FROM unitemp_coalesce_select_effective(
    'SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted'::text, 
    ARRAY['emp_name', 'dept']
    ) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;

SELECT * FROM unitemp_coalesce_table_effective(
    'public', 'employee', ARRAY['emp_name', 'dept'])
AS (emp_name text,
        dept text,
        effective tstzrange) ORDER BY emp_name, effective;

SELECT * FROM v8_select_test('dsadsa', '0.5 year'::interval, '{}');

SELECT MAX(UPPER(effective)), date_trunc('years', MAX(UPPER(effective))) FROM employee WHERE UPPER(effective) != 'infinity' AND now() <@ asserted;

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'employee2',
	$$emp_id int, 
	  emp_name text not null,
    dept text not null
	$$,
   'emp_id');

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$1, 'George', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-08'::timestamptz, '2010-10-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$2, 'Ron', 'Ship'$$,
    temporal_relationships.timeperiod('2010-09-28'::timestamptz, '2010-10-01'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$3, 'Bob', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-10'::timestamptz, '2010-10-15'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);

SELECT * FROM unitemp_coalesce_select_effective(
    $$
    SELECT * FROM (
    SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted
    UNION
    SELECT emp_name, dept, effective FROM employee2 WHERE now() <@ asserted
    ) t
    $$,
    ARRAY['emp_name', 'dept']
) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;

SELECT * FROM unitemp_coalesce_select(
    $$
    SELECT * FROM (
    SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted
    UNION
    SELECT emp_name, dept, effective FROM employee2 WHERE now() <@ asserted
    ) t
    $$,
    ARRAY['emp_name', 'dept'],
    'effective'
) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;


DROP TABLE employee;
DROP TABLE employee2;
DROP TABLE manages;