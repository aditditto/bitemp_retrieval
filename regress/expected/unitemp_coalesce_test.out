DROP EXTENSION bitemp_retrieval;
CREATE EXTENSION bitemp_retrieval;
SET DATESTYLE TO 'ISO';
SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'employee',
	$$emp_id int, 
	  emp_name text not null,
    dept text not null
	$$,
   'emp_id');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$1, 'Ron', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-01'::timestamptz, '2010-10-06'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$2, 'George', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-05'::timestamptz, '2010-10-09'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee',
    $$emp_id, emp_name, dept$$,
    $$1, 'Ron', 'Mail'$$,
    temporal_relationships.timeperiod('2010-10-06'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Mail'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-08'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Ship'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-10'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_update(
    'public',
    'employee',
    $$emp_name, dept$$,
    $$'Ron', 'Ship'$$,
    'emp_id',
    $$1$$,
    temporal_relationships.timeperiod('2010-10-15'::timestamptz, 'infinity'), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_update 
----------------------
                    1
(1 row)

SELECT emp_id, emp_name, dept, effective FROM employee;
 emp_id | emp_name | dept |                      effective                      
--------+----------+------+-----------------------------------------------------
      1 | Ron      | Ship | ["2010-10-01 00:00:00-07","2010-10-06 00:00:00-07")
      2 | George   | Ship | ["2010-10-05 00:00:00-07","2010-10-09 00:00:00-07")
      1 | Ron      | Mail | ["2010-10-06 00:00:00-07",infinity)
      1 | Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-08 00:00:00-07")
      1 | Ron      | Mail | ["2010-10-08 00:00:00-07",infinity)
      1 | Ron      | Mail | ["2010-10-08 00:00:00-07","2010-10-10 00:00:00-07")
      1 | Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
      1 | Ron      | Ship | ["2010-10-10 00:00:00-07","2010-10-15 00:00:00-07")
      1 | Ron      | Ship | ["2010-10-15 00:00:00-07",infinity)
(9 rows)

SELECT emp_id, emp_name, dept, effective FROM employee WHERE now() <@ asserted ORDER BY emp_name, asserted;
 emp_id | emp_name | dept |                      effective                      
--------+----------+------+-----------------------------------------------------
      2 | George   | Ship | ["2010-10-05 00:00:00-07","2010-10-09 00:00:00-07")
      1 | Ron      | Ship | ["2010-10-01 00:00:00-07","2010-10-06 00:00:00-07")
      1 | Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-08 00:00:00-07")
      1 | Ron      | Mail | ["2010-10-08 00:00:00-07","2010-10-10 00:00:00-07")
      1 | Ron      | Ship | ["2010-10-10 00:00:00-07","2010-10-15 00:00:00-07")
      1 | Ron      | Ship | ["2010-10-15 00:00:00-07",infinity)
(6 rows)

SELECT * FROM unitemp_coalesce_select(
    'SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted'::text, 
    ARRAY['emp_name', 'dept'], 
    'effective'::text
    ) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;
 emp_name | dept |                      effective                      
----------+------+-----------------------------------------------------
 George   | Ship | ["2010-10-05 00:00:00-07","2010-10-09 00:00:00-07")
 Ron      | Ship | ["2010-10-01 00:00:00-07","2010-10-06 00:00:00-07")
 Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-10 00:00:00-07")
 Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
(4 rows)

SELECT * FROM unitemp_coalesce_select_effective(
    'SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted'::text, 
    ARRAY['emp_name', 'dept']
    ) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;
 emp_name | dept |                      effective                      
----------+------+-----------------------------------------------------
 George   | Ship | ["2010-10-05 00:00:00-07","2010-10-09 00:00:00-07")
 Ron      | Ship | ["2010-10-01 00:00:00-07","2010-10-06 00:00:00-07")
 Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-10 00:00:00-07")
 Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
(4 rows)

SELECT * FROM unitemp_coalesce_table_effective(
    'public', 'employee', ARRAY['emp_name', 'dept'])
AS (emp_name text,
        dept text,
        effective tstzrange) ORDER BY emp_name, effective;
 emp_name | dept |                      effective                      
----------+------+-----------------------------------------------------
 George   | Ship | ["2010-10-05 00:00:00-07","2010-10-09 00:00:00-07")
 Ron      | Ship | ["2010-10-01 00:00:00-07","2010-10-06 00:00:00-07")
 Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-10 00:00:00-07")
 Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
(4 rows)

SELECT * FROM v8_select_test('dsadsa', '0.5 year'::interval, '{}');
INFO:  NaN
INFO:  NaN
INFO:  false
INFO:  false
INFO:  [-infinity,infinity]
INFO:  string
INFO:  15552000
INFO:  @ 6 mons
INFO:  string
INFO:  0
 v8_select_test 
----------------
 
(1 row)

SELECT MAX(UPPER(effective)), date_trunc('years', MAX(UPPER(effective))) FROM employee WHERE UPPER(effective) != 'infinity' AND now() <@ asserted;
          max           |       date_trunc       
------------------------+------------------------
 2010-10-15 00:00:00-07 | 2010-01-01 00:00:00-08
(1 row)

SELECT * FROM bitemporal_internal.ll_create_bitemporal_table(
    'public',
    'employee2',
	$$emp_id int, 
	  emp_name text not null,
    dept text not null
	$$,
   'emp_id');
 ll_create_bitemporal_table 
----------------------------
 t
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$1, 'George', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-08'::timestamptz, '2010-10-11'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$2, 'Ron', 'Ship'$$,
    temporal_relationships.timeperiod('2010-09-28'::timestamptz, '2010-10-01'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT * FROM bitemporal_internal.ll_bitemporal_insert(
    'public.employee2',
    $$emp_id, emp_name, dept$$,
    $$3, 'Bob', 'Ship'$$,
    temporal_relationships.timeperiod('2010-10-10'::timestamptz, '2010-10-15'::timestamptz), --effective
    temporal_relationships.timeperiod(now(), 'infinity') --asserted
);
 ll_bitemporal_insert 
----------------------
                    1
(1 row)

SELECT emp_id, emp_name, dept, effective FROM employee2 WHERE now() <@ asserted ORDER BY emp_name, asserted;
 emp_id | emp_name | dept |                      effective                      
--------+----------+------+-----------------------------------------------------
      3 | Bob      | Ship | ["2010-10-10 00:00:00-07","2010-10-15 00:00:00-07")
      1 | George   | Ship | ["2010-10-08 00:00:00-07","2010-10-11 00:00:00-07")
      2 | Ron      | Ship | ["2010-09-28 00:00:00-07","2010-10-01 00:00:00-07")
(3 rows)

SELECT * FROM unitemp_coalesce_select_effective(
    $$
    SELECT * FROM (
    SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted
    UNION
    SELECT emp_name, dept, effective FROM employee2 WHERE now() <@ asserted
    ) t
    $$,
    ARRAY['emp_name', 'dept']
) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;
 emp_name | dept |                      effective                      
----------+------+-----------------------------------------------------
 Bob      | Ship | ["2010-10-10 00:00:00-07","2010-10-15 00:00:00-07")
 George   | Ship | ["2010-10-05 00:00:00-07","2010-10-11 00:00:00-07")
 Ron      | Ship | ["2010-09-28 00:00:00-07","2010-10-06 00:00:00-07")
 Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-10 00:00:00-07")
 Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
(5 rows)

SELECT * FROM unitemp_coalesce_select(
    $$
    SELECT * FROM (
    SELECT emp_name, dept, effective FROM employee WHERE now() <@ asserted
    UNION
    SELECT emp_name, dept, effective FROM employee2 WHERE now() <@ asserted
    ) t
    $$,
    ARRAY['emp_name', 'dept'],
    'effective'
) AS (emp_name text,
            dept text,
            effective tstzrange) ORDER BY emp_name, effective;
 emp_name | dept |                      effective                      
----------+------+-----------------------------------------------------
 Bob      | Ship | ["2010-10-10 00:00:00-07","2010-10-15 00:00:00-07")
 George   | Ship | ["2010-10-05 00:00:00-07","2010-10-11 00:00:00-07")
 Ron      | Ship | ["2010-09-28 00:00:00-07","2010-10-06 00:00:00-07")
 Ron      | Mail | ["2010-10-06 00:00:00-07","2010-10-10 00:00:00-07")
 Ron      | Ship | ["2010-10-10 00:00:00-07",infinity)
(5 rows)

DROP TABLE employee;
DROP TABLE employee2;
DROP TABLE manages;
ERROR:  table "manages" does not exist
